domain/network/state/types.go

```diff
@@ -0,0 +1,236 @@
+
+package state
+
+import (
+	"database/sql"
+
+	"github.com/juju/juju/core/network"
+)
+
+// Space represents a single row from the database when
+// space is joined with provider_space, subnet, subnet_type,
+// subnet_association, subject_subnet_type_uuid, subnet_association_type,
+// provider_subnet, provider_network, provider_network_subnet,
+// availability_zone and availability_zone_subnet.
+type Space struct {
+	// UUID is the space UUID.
+	UUID string `db:"uuid"`
+
+	// Name is the space name.
+	Name string `db:"name"`
+
+	// ProviderID is the space provider id.
+	ProviderID sql.NullString `db:"provider_id"`
+
+	// SubnetUUID is one of the space's subnet id.
+	SubnetUUID string `db:"subnet_uuid"`
+
+	// CIDR is one of the space's subnet cidr.
+	CIDR string `db:"subnet_cidr"`
+
+	// VLANTag is one of the space's subnet vlan tag.
+	VLANTag int `db:"vlan_tag"`
+
+	// SubnetProviderID is one of the space's subnet provider id.
+	SubnetProviderID string `db:"subnet_provider_id"`
+
+	// SubnetProviderNetworkID is one of the space's subnet provider network id.
+	SubnetProviderNetworkID string `db:"subnet_provider_network_id"`
+
+	// SubnetAZ is one of the availability zones on one of the subnets of the space.
+	SubnetAZ string `db:"subnet_az"`
+
+	// SubnetOverlayCIDR is one of the space's subnet overlay cidr in a fan setup.
+	SubnetOverlayCIDR sql.NullString `db:"subnet_overlay_cidr"`
+
+	// SubnetUnderlayCIDR is one of the space's subnet underlay cidr in a fan setup.
+	SubnetUnderlayCIDR sql.NullString `db:"subnet_underlay_cidr"`
+}
+
+// Alias type to a slice of Space('s).
+type Spaces []Space
+
+// ToSpaceInfos converts Spaces to a slice of network.SpaceInfo structs.
+// This method makes sure only unique subnets are mapped and flattens them into
+// each space.
+// No sorting is applied.
+func (sp Spaces) ToSpaceInfos() network.SpaceInfos {
+	var res network.SpaceInfos
+
+	// Prepare structs for unique subnets for each space.
+	uniqueAZs := make(map[string]map[string]map[string]string)
+	uniqueSubnets := make(map[string]map[string]network.SubnetInfo)
+	uniqueSpaces := make(map[string]network.SpaceInfo)
+
+	for _, space := range sp {
+		spInfo := network.SpaceInfo{
+			ID:   space.UUID,
+			Name: network.SpaceName(space.Name),
+		}
+
+		if _, ok := uniqueSubnets[space.UUID]; !ok {
+			uniqueSubnets[space.UUID] = make(map[string]network.SubnetInfo)
+		}
+		snInfo := network.SubnetInfo{
+			ID:                network.Id(space.SubnetUUID),
+			CIDR:              space.CIDR,
+			ProviderId:        network.Id(space.SubnetProviderID),
+			ProviderNetworkId: network.Id(space.SubnetProviderNetworkID),
+			VLANTag:           space.VLANTag,
+			SpaceID:           space.UUID,
+			SpaceName:         space.Name,
+		}
+		if space.ProviderID.Valid {
+			spInfo.ProviderId = network.Id(space.ProviderID.String)
+			snInfo.ProviderSpaceId = network.Id(space.ProviderID.String)
+		}
+		if space.SubnetOverlayCIDR.Valid || space.SubnetUnderlayCIDR.Valid {
```

Comment:
Overlay is never valid, because you never select into it. This has to be the /8 derived from the CIDR or a subnet with type overlay anyway.

Reply:
sure, fixed in ef53aca596